theano/sparse/basic.py

```diff
@@ -1346,6 +1346,23 @@ def perform(self, node, inputs, outputs):
         assert _is_sparse(x)
         out[0] = x[start1:stop1:step1, start2:stop2:step2]
 
+    def grad(self, inputs, grads):
+        x = inputs[0]
+        rest = inputs[1:]
+        gz, = grads
+
+        assert _is_sparse_variable(x) == True
+
+        return [x.zeros_like() + gz] + [DisconnectedType()()] * len(rest)
```

Comment:
gz and x.zeros_like() don't have the same shape. gz have the shape of the output of the forword of this op.

Doing the correct computation will request a new op that take 3 inputs:

x.zeros_like(), gz and index.

Then it will have a perform method that will do:
x[index] += gz

or

x[index] = gz

Also, check in the doc, they tell how to test the grad. It will help you know when it is right. If you have questions, don't hesitate.